src/Api/FormatsProvider.php

```diff
@@ -0,0 +1,76 @@
+<?php
+
+/*
+ * This file is part of the API Platform project.
+ *
+ * (c) Kévin Dunglas <dunglas@gmail.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace ApiPlatform\Core\Api;
+
+use ApiPlatform\Core\Exception\InvalidArgumentException;
+use ApiPlatform\Core\Metadata\Resource\Factory\ResourceMetadataFactoryInterface;
+
+/**
+ * {@inheritdoc}
+ *
+ * @author Anthony GRASSIOT <antograssiot@free.fr>
+ */
+final class FormatsProvider implements FormatsProviderInterface
+{
+    private $configuredFormats;
+    private $resourceMetadataFactory;
+
+    public function __construct(ResourceMetadataFactoryInterface $resourceMetadataFactory, array $configuredFormats)
+    {
+        $this->resourceMetadataFactory = $resourceMetadataFactory;
+        $this->configuredFormats = $configuredFormats;
+    }
+
+    /**
+     * Finds formats from a Resource class.
+     */
+    public function getFormatsFromAttributes(array $attributes): array
+    {
+        if (!$attributes || !isset($attributes['resource_class'])) {
+            return $this->configuredFormats;
+        }
+
+        $resourceMetadata = $this->resourceMetadataFactory->create($attributes['resource_class']);
+
+        if (!$formats = (array) $resourceMetadata->getOperationAttribute($attributes, 'formats', [], true)) {
+            return $this->configuredFormats;
+        }
+
+        return $this->getOperationFormats($formats);
+    }
+
+    /**
+     * Filter and populate the acceptable formats.
+     *
+     * @throws InvalidArgumentException
+     */
+    private function getOperationFormats(array $annotationFormats): array
+    {
+        $resourceFormats = [];
+        foreach ($annotationFormats as $format => $value) {
+            if (!is_numeric($format)) {
+                $resourceFormats[$format] = (array) $value;
+                continue;
+            }
+            if (array_key_exists($value, $this->configuredFormats)) {
```

Comment:
According to the main feature of this PR, `$value` could be an array (`"formats"={"json", "pdf"={"application/pdf"}}`). The method `array_key_exists` does not allow an array for the first property, and PHP will throw a warning:
![array_key_exists](https://user-images.githubusercontent.com/407859/38452879-1ce91532-3a4d-11e8-9d78-ec454563ab9a.png)

Reply:
The `if` statement may return false if the first argument of `array_key_exists` isn't numeric or a string. I recommend you to check `$value` type before calling `array_key_exists`

Reply:
Indeed @vincentchalamon, I failled to see that, thanks for your time and explanation, I've added the check and a new exception message, I'm not found on the wording though, you might have a better idea...